### PR TITLE
new network feature: NAPT (widely known as NAT)

### PR DESCRIPTION
**Pull request by d-a-v**
This is brought to us thanks to
- @NeoCat ([branch](https://github.com/NeoCat/esp8266-Arduino/tree/ip_napt), [commit](https://github.com/NeoCat/esp8266-Arduino/commit/4108c8dbced7769c75bcbb9ed880f1d3f178bcbe)) who made this a while ago for lwIP-1.4
- @jonask1337 and @martin-ger ([patches](https://github.com/jonask1337/esp-lwip)) who ported the above to esp-idf / lwIP-2.0.3

My contribution was easy ([ref](https://github.com/martin-ger/esp_wifi_repeater/issues/333#issuecomment-513718331)), I take no credit.
Napt is now ported to lwIP-2.1.2 (latest stable) as a real patch in lwip2 so it can be proposed upstream.

It is currently only enabled with lwip2 (no IPv6, with features).

This PR also brings heap optimizations from @earlephilhower and improved locking macros from @mhightower83 .

An example is proposed (filling half the RAM) for a damn simple and basic WiFi range extender.

### CODE REVIEW

File: libraries/ESP8266WiFi/examples/RangeExtender-NAPT/RangeExtender-NAPT.ino [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+#include <lwip/napt.h>
+#include <lwip/dns.h>
+#include <dhcpserver.h>
+
+#define NAPT 1000
+#define NAPT_PORT 10
+
+#if HAVE_NETDUMP
+
+#include <NetDump.h>
```
[the remark this answers is not in the document]
Comment (d-a-v) [reply libraries/ESP8266WiFi/examples/RangeExtender-NAPT/RangeExtender-NAPT.ino]: TBH, I'd like to have it in the core. 
@hreintke is willing to propose us an improved version of NetDump library so I don't push mine (see https://github.com/d-a-v/EspGoodies/pull/2)

I think it is nice to let this here as-is so users who want to test NAT before using it have the option to check with this library (currently mine - I may add the URL, but the example can be later updated when we can import @hreintke's version right in this repo)

File: libraries/ESP8266WiFi/examples/RangeExtender-NAPT/RangeExtender-NAPT.ino [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    delay(500);
+  }
+  Serial.printf("\nSTA: %s (dns: %s / %s)\n",
+                WiFi.localIP().toString().c_str(),
+                WiFi.dnsIP(0).toString().c_str(),
+                WiFi.dnsIP(1).toString().c_str());
+
+  // give DNS servers to AP side
+  dhcps_set_dns(0, WiFi.dnsIP(0));
+  dhcps_set_dns(1, WiFi.dnsIP(1));
```
[the remark this answers is not in the document]
Comment (d-a-v) [reply libraries/ESP8266WiFi/examples/RangeExtender-NAPT/RangeExtender-NAPT.ino]: > This dns stuff should be wrapped somehow in a method in the WiFi class.

True. An API has to be defined.

File: libraries/ESP8266WiFi/examples/RangeExtender-NAPT/RangeExtender-NAPT.ino [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+  WiFi.softAPConfig(  // enable AP, with android-compatible google domain
+    IPAddress(172, 217, 28, 254),
+    IPAddress(172, 217, 28, 254),
+    IPAddress(255, 255, 255, 0));
+  WiFi.softAP(STASSID "extender", STAPSK);
+  Serial.printf("AP: %s\n", WiFi.softAPIP().toString().c_str());
+
+  Serial.printf("Heap before: %d\n", ESP.getFreeHeap());
+  err_t ret = ip_napt_init(NAPT, NAPT_PORT);
```
[the remark this answers is not in the document]
Comment (d-a-v) [reply libraries/ESP8266WiFi/examples/RangeExtender-NAPT/RangeExtender-NAPT.ino]: > These napt lower level calls likely should also be wrapped in a method in the wifi class.

True. An API has to be defined.

File: libraries/ESP8266WiFi/examples/RangeExtender-NAPT/RangeExtender-NAPT.ino [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  Serial.printf("ip_napt_init(%d,%d): ret=%d (OK=%d)\n", NAPT, NAPT_PORT, (int)ret, (int)ERR_OK);
+  if (ret == ERR_OK) {
+    ret = ip_napt_enable_no(SOFTAP_IF, 1);
```
[the remark this answers is not in the document]
Comment (d-a-v) [reply libraries/ESP8266WiFi/examples/RangeExtender-NAPT/RangeExtender-NAPT.ino]: `!0` is `enable`
It should also be wrapped in an API anyway.

**Comment by Suraj151** [comment]
### Platform

- Hardware: ESP-12E
- Core Version: dont know following doc https://arduino--esp8266.readthedocs.io/en/latest/installing.html#instructions-other-os
- Development Env: Arduino IDE 1.8.7
- Operating System: Ubuntu

### Settings in IDE

- Module: Nodemcu 1.0
- Flash Mode: not listed
- Flash Size: 4MB
- lwip Variant: v2 Lower Memory
- CPU Frequency: 80Mhz
- Upload Using: SERIAL
- Upload Speed: 115200

### Problem Description

i tried napt example with fresh arduino installation but it didn't work as it work with lwip1.4. end device not get internet access of router. i tested it with range extender example.

### Debug Messages

```
NAPT Range extender
Heap on start: 50112
...........
STA: 192.168.137.2 (dns: 192.168.137.1 / (IP unset))
AP: 192.168.0.1
Heap before: 49400
ip_napt_init(1000,10): ret=0 (OK=0)
ip_napt_enable_no(SOFTAP_IF): ret=0 (OK=0)
WiFi Network 'Windows Phoneextender' with same password is now NATed behind 'Windows Phone'
Heap after napt init: 36544
```

is there any other setting that i missed ?

**Comment by Suraj151** [comment]
### it works !

i dont know how but when i reinstalled all things with same steps in another pc. then it works as expected.
will try to understand reason and thanks for such awaited feature.

**Pull request merged**